File: devices/AquaWiiRemote/AquaWiiRemote/UnifiedZoomEvent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace AquaWiiRemote
{
    class UnifiedZoomEvent : UnifiedEvent
    {
        private float _scale;
        private float[] _center = new float[3];

        public UnifiedZoomEvent(String name, String desc, byte type, int id,
                float[] location, float scale, float[] center)
            : base(name, desc, type, id, location)
        {
            _scale = scale;
            _center[0] = center[0];
            _center[1] = center[1];
            _center[2] = center[2];
        }

        protected override byte[] serializeData(out short outLength)
        {
            outLength = 16;
            byte[] ret = new byte[16];
            float tempScale, tempX, tempY, tempZ;
            byte[] temp = new byte[4];
            int index = 0;

            tempScale = _scale;
            tempX = _center[0];
            tempY = _center[1];
            tempZ = _center[2];

            // Scale
            temp = System.BitConverter.GetBytes(_scale);
            endianSwap(ref temp);
            Array.Copy(temp, 0, ret, index, temp.Length);
            index += temp.Length;

            // Location
            for (int i = 0; i < 3; i++)
            {
                temp = System.BitConverter.GetBytes(_center[i]);
                endianSwap(ref temp);
                Array.Copy(temp, 0, ret, index, temp.Length);
                index += temp.Length;
            }

            return ret;
            
        }
    }
}

File: trunk/devices/AquaWinMouse/UnifiedZoomEvent.cpp
/**
 * UnifiedZoomEvent.cpp
 *
 * An event to use for sending zoom-related information.
 * 
 * This file is part of Aqua Universal Gesture Recognition Framework.
 * Copyright (C) 2010  Jay Roltgen.
 *
 * Aqua is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "UnifiedZoomEvent.h"

#include <stdio.h>

using namespace std;

UnifiedZoomEvent::UnifiedZoomEvent(char *data) : Event(data) {
    int i;
    int dataPos = (_name.length() + _description.length() + 2 + 17);
    
    memcpy(&_zoomScale, &data[dataPos], 4);
    dataPos += 4;
    
    memcpy(_zoomCenter, &data[dataPos], 12);
    
    // Handle endianness.
    if (EndianConverter::isLittleEndian()) {
        _zoomScale = EndianConverter::swapFloatEndian(_zoomScale);
        for (int i = 0; i < 3; i++) {
            _zoomCenter[i] = EndianConverter::swapFloatEndian(_zoomCenter[i]);
        }
    }
}

UnifiedZoomEvent::UnifiedZoomEvent(string& name, string& desc, 
        char type, int id, float* location, float scale, float* center) : 
        Event(name, desc, type, id, location) {
    
    _zoomScale = scale;
    memcpy(_zoomCenter, center, 12);    
}
            
float UnifiedZoomEvent::getScale() {
    return _zoomScale;
}

float* UnifiedZoomEvent::getCenter() {
    return _zoomCenter;
}
    
/**
 * Constructs a char array with this event's data.  Data:
 *  - 4 bytes : zoom scale (float)
 *  - 4 bytes : zoom center x-coord
 *  - 4 bytes : zoom center y-coord
 *  - 4 bytes : zoom center z-coord
 */
char* UnifiedZoomEvent::serializeData(short& outLength) {
    
    outLength = UNIFIEDZOOMEVENT_DATA_LENGTH;
    
    float tempScale, tempX, tempY, tempZ;
    tempScale = _zoomScale;
    // zoom center
    tempX = _zoomCenter[0];
    tempY = _zoomCenter[1];
    tempZ = _zoomCenter[2];
    
    if (EndianConverter::isLittleEndian()) {
        tempScale = EndianConverter::swapFloatEndian(tempScale);
        tempX = EndianConverter::swapFloatEndian(tempX);
        tempY = EndianConverter::swapFloatEndian(tempY);
        tempZ = EndianConverter::swapFloatEndian(tempZ);
    }
    memcpy(_dataBuffer + 0, &tempScale, 4);
    memcpy(_dataBuffer + 4, &tempX, 4);
    memcpy(_dataBuffer + 8, &tempY, 4);
    memcpy(_dataBuffer + 12, &tempZ, 4);
    
    return _dataBuffer;

}